fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/GameLifecycleCallback.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.event.lifecycle.v1;
+
+/**
+ * A callback which is invoked when the game's engine advances in it's lifecycle.
+ */
+public interface GameLifecycleCallback<T> {
```

Comment:
I'd use two instances without generics for ease of use. It may be worthwhile to organize the events+callbacks differently, maybe an outer class like ClientTickEvents that holds both the event fields and callbacks as nested classes? Similar for others, ClientLifecycleEvents is getting big and the many trivial top level callback interfaces are getting a bit too spread out.

Reply:
I've split up the `Client`/`ServerLifecycleEvents` into more specific classes.

Because of the split, at most these more specific classes (So there is a `ClientTickEvents.World` and `ServerTickEvents.World`) now only hold 2-3 inner classes for the interfaces for event listeners implement.